src/dom/JsValueConversion.ts

```diff
@@ -0,0 +1,142 @@
+            [IonTypes.LIST, dom.List],
+            [IonTypes.STRUCT, dom.Struct],
+        ]);
+    }
+    return _domTypesByIonType;
+}
+
+// Returns a dom type constructor for the provided IonType.
+// This function is exported to assist in unit testing but is not visible at the library level.
+export function _domConstructorFor(ionType: IonType): Constructor {
+    let domConstructor = _getDomTypesByIonTypeMap().get(ionType)!;
+    if (!_hasValue(domConstructor)) {
+        throw new Error(`No dom type constructor was found for Ion type ${ionType.name}`);
+    }
+    return domConstructor;
+}
+
+// Examines the provided JS value and returns the IonType best suited to represent it.
+// This function will never return Clob, SExpression, and Symbol; it will always
+// favor Blob, List, and String respectively.
+function _inferType(value: any): IonType {
+    if (value === undefined) {
+        throw new Error("Cannot create an Ion value from `undefined`.");
+    }
+    if (value === null) {
+        return IonTypes.NULL;
+    }
+
+    let valueType: string = typeof(value);
+    switch (valueType) {
+        case "string":
+            return IonTypes.STRING;
+        case "number":
+            return Number.isInteger(value)
+                ? IonTypes.INT
+                : IonTypes.FLOAT;
+        case "boolean":
+            return IonTypes.BOOL;
+        case "object":
+            break;
+        // We `break` here to defer handling native bigints to the 'valueof JSBI' test below,
+        // which will handle native bigints correctly if Babel is being used to convert JSBI
+        // logic to native bigint logic.
+        case "bigint":
+            break;
+
+        //
+        // TODO: Javascript symbols are not a 1:1 match for Ion symbols, but we may wish
+        //       to support them in Value.from() anyway.
+        case "symbol":
+        default:
+            throw new Error(`Value.from() does not support the JS primitive type ${valueType}.`);
+    }
+
+    // If `value` is a native bigint but we're not using Babel for native bigint support, we need to
+    // throw an Error.
+    //
+    // If Babel is being used, this test becomes:
+    //      typeof(value) === 'bigint'
+    // which will return true for native bigints, leading to the creation of an Integer.
+    //
+    // If Babel is not being used, then this remains a JSBI instanceof test, which will
+    // return false for native bigints.
+    if (value instanceof JSBI) {
+        return IonTypes.INT;
+    } else if (typeof(value) === 'bigint') {
+        // If Babel is being used, this check will be identical to the one above it and it will
+        // never be reached. If Babel is not being used, this test will return true for native
+        // bigints (which aren't supported without Babel) while allowing other Object types
+        // to be processed below.
+        throw new Error('bigints are not supported without using Babel for JSBI compilation.');
+    }
```

Comment:
This code is a bit mind-bending, unfortunately. It needs to work regardless of whether Babel/JSBI is being used, which makes it trickier to understand--hopefully the comments help.

Reply:
oof.  the comments certainly help.  problematic if someone depends on the ES5 build, but at some point their app starts using bigints--at which point i think we'd encourage them to switch to the ES6 build

Reply:
> oof. the comments certainly help. problematic if someone depends on the ES5 build, but at some point their app starts using bigints--at which point i think we'd encourage them to switch to the ES6 build

I've added a `TODO`--I think we can use this approach to create an easier-to-understand `_babelForJsbiIsEnabled()` test.